Automattic/jetpack
# Unbundle Social JS code from Jetpack and Social

Currently, the JS code of social that gets bundled in Jetpack takes precedence over Social even if Social is the latest version. This makes it difficult to fix a bug in Jetpack given the monthly release cycle. Thus we should explore ways to ensure that the JS code is not dependent upon what has been bundled with the plugin. The latest version of the JS code should be used if it's present in any of the plugins.

P2 post: pdWQjU-Mc-p2

This PR is a proof of concept to test unbundling of publicize and the related JS code/logic from both Jetpack and Social plugin. This moves the JS bundling from Jetpack/Social to the publicize package to ensure that the latest version of the code is always loaded regardless of which plugin is active.

## Proposed changes:

* Create webpack utils for dependency extraction. These utilities can be used by other plugins if they need to.
* Prepare publicize package for bundling by adding a webpack config
* Update a few components in Jetpack and Social for testing the PoC.
* To ensure that the separate bundling works correctly, we need to ensure that the initial state is rendered before the publicize handle and thus this PR changes the handles for initial state.

### Approach/idea
The basic idea is to make Jetpack independent of `js-packages/publicize-components` package and rather depend upon `publicize` package. So, if Jetpack or Social needs to render something common from `publicize-components`, it must be exported via publicize package. The Social specific code like the admin page may remain in Social plugin itself, i.e. it may consume `publicize-components`.

The publicize package will export the shared components via its own bundle on runtime just like WP core components and other packages.

This will ensure that we will always have the latest version of publicize UI both in Jetpack and Social.

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






* Ensure that there is no `jetpack watch ...` command running for any plugin
* Install the dependencies - `pnpm install`
* Run `jepack build plugins/social` and `jetpack build plugins/jetpack`
* Now run `jetpack build packages/publicize`
* Now goto connections management on both Jetpack and Social settings pages
* Confirm that everything is as before
* Now make a change to to a file used in connections management, e.g. change the button label from `'Connect an account'` to something else in `publicize-components/src/components/connection-management/index.tsx`, e.g. `'Connect a social account'`
* Now build only publicize package - `jetpack build packages/publicize`
* Goto connections management again on both Jetpack and Social settings pages
* Confirm that the button label is what you updated to.
* Now change that button label again to something different.
* Now build the plugins not the package - `jepack build plugins/social` and `jetpack build plugins/jetpack`
* Confirm that the label is not changes because publicize wasn't built yet.
* For simple sites, apply D155515-code and this PR to your sandbox and point your site and public API to your sandbox.
* Goto post editor on that simple site
* Confirm that Jetpack sidebar loads correctly and everything there works as before.
